src/loader/bgw_launcher.c

```diff
@@ -588,25 +588,55 @@ launcher_handle_message(HTAB *db_htab)
 	return true;
 }
 
+/* ordered by priority, SIGTERM should take precedence */
+typedef enum ShutdownSignal
+{
+	NO_SIGNAL = 0,
+	GOT_SIGINT,
+	GOT_SIGTERM,
+}			ShutdownSignal;
+
+static volatile sig_atomic_t shutdown_signal = NO_SIGNAL;
 static void
 bgw_sigterm(SIGNAL_ARGS)
 {
-	/*
-	 * The launcher should never be in a critical section, but if it is we
-	 * have no choice but to PANIC since it is unsafe to exit.
-	 */
-	if (CritSectionCount != 0)
-		ereport(PANIC, (errmsg("TimescaleDB background worker launcher terminated while in a critical section")));
-	ereport(LOG, (errmsg("TimescaleDB background worker launcher terminated by administrator command. Launcher will not restart after exiting")));
-	proc_exit(0);
+	shutdown_signal = GOT_SIGTERM;
+	SetLatch(MyLatch);
 }
 
 static void
 bgw_sigint(SIGNAL_ARGS)
 {
-	ereport(ERROR, (errmsg("TimescaleDB background worker launcher canceled by administrator command. Launcher will restart after exiting")));
+	if (shutdown_signal < GOT_SIGINT)
+		shutdown_signal = GOT_SIGINT;
+	SetLatch(MyLatch);
 }
 
+static void
+handle_shutdown_signal()
+{
+
+	switch (shutdown_signal)
+	{
+		case GOT_SIGTERM:
+
+			/*
+			 * The launcher should never be in a critical section, but if it
+			 * is we have no choice but to PANIC since it is unsafe to exit.
+			 */
+			if (CritSectionCount != 0)
+				ereport(PANIC, (errmsg("TimescaleDB background worker launcher terminated while in a critical section")));
+			ereport(LOG, (errmsg("TimescaleDB background worker launcher terminated by administrator command. Launcher will not restart after exiting")));
+			proc_exit(0);
```

Comment:
Why do we explicitly exit here as opposed to just returning from the main function? I think breaking out of the main loop is what you would normally do.  IMO opinion, the code is harder to understand if we exit at arbitrary places rather than just executing the main function to its end, especially in case you do things like cleanup, etc.